### ISSUE TITLE
Obtaining flows from the Policy Object

### ISSUE THREAD

**Issue by nopbyte** [SEDARI/UPFROnt#5]
Hi @dschrecking 

I am using ulocks@v0.5.1 and upfront@v0.4.2

My plan was to pass the policy that is returned by the PAP to the upfront.policy object (used as a constructor) and then execute getSimpleFlows. But I cannot build the Policy object.  

After reading the policy from the PAP and passing it to upfront.policy I am getting this:
```Error: Policy: Cannot construct rule from flows which are not contained in an array!```

My code is doing
```
var Policy = ulocks.Policy;
var p = new Policy(policy);
```

where policy contains the following object:

```{
	"d": {
		"0": {
			"p": [{
				"op": "write"
			}, {
				"op": "read"
			}],
			"c": 1
		},
		"1": {
			"p": {
				"flows": [{
					"op": "read"
				}, {
					"op": "write",
					"locks": [{
						"lock": "hasType",
						"args": ["/user"]
					}, {
						"lock": "isOwner"
					}]
				}, {
					"op": "write",
					"locks": [{
						"lock": "hasType",
						"args": ["/user"]
					}, {
						"lock": "attrEq",
						"args": ["role", "admin"]
					}]
				}],
				"actions": {
					"read": [{
						"action": "delete"
					}]
				}
			},
			"c": 3
		},
		"2": {
			"p": [{
				"op": "read",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "isOwner"
				}]
			}, {
				"op": "read",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "attrEq",
					"args": ["role", "admin"]
				}]
			}, {
				"op": "write",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "isOwner"
				}]
			}, {
				"op": "write",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "attrEq",
					"args": ["role", "admin"]
				}]
			}],
			"c": 1
		},
		"3": {
			"p": [{
				"op": "read"
			}, {
				"op": "write",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "isOwner"
				}]
			}, {
				"op": "write",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "attrEq",
					"args": ["role", "admin"]
				}]
			}],
			"c": 1
		},
		"4": {
			"p": [{
				"op": "read"
			}],
			"c": 1
		},
		"5": {
			"p": [{
				"op": "read"
			}, {
				"op": "write",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "isOwner"
				}]
			}],
			"c": 1
		},
		"6": {
			"p": [{
				"op": "read",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "isOwner"
				}]
			}, {
				"op": "write",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "isOwner"
				}]
			}, {
				"op": "write",
				"locks": [{
					"lock": "hasType",
					"args": ["/user"]
				}, {
					"lock": "attrEq",
					"args": ["role", "admin"]
				}]
			}],
			"c": 1
		}
	},
	"e": 0,
	"o": {
		"s": 1,
		"m": null,
		"p": {
			"actions": {
				"s": 2,
				"m": null,
				"p": {}
			},
			"policies": {
				"s": 3,
				"m": null,
				"p": {
					"policies": {
						"s": null,
						"m": null,
						"p": {
							"policies": {
								"s": 4,
								"m": null,
								"p": {}
							}
						}
					}
				}
			},
			"groups": {
				"s": 1,
				"m": null,
				"p": {}
			},
			"entities": {
				"s": 1,
				"m": null,
				"p": {}
			},
			"credentials": {
				"s": 5,
				"m": null,
				"p": {
					"dropbox": {
						"s": 6,
						"m": null,
						"p": {}
					}
				}
			}
		}
	}
}```

Thanks a lot for your help

**Comment by dschrecking** [comment]
Hi @nopbyte ,

`policy` is not a ULock policy. This is a policyObject used in the PAP. You usually retrieve these type of objects by the PAP call getFullRecord. A policy however, can only be retrieved if you actually say for which property you want the policy or by retrieving a map which generates a property to policy mapping for the full object. 

How did you get the 'policy' you show above?

Cheers

**Comment by nopbyte** [comment]
Hi @dschrecking 

Yes I am getting this object from the getFullRecord. This is what I was using before to check if a policy was set properly. 

I have seen that I could use getAllProperties to get the list of properties set for a policy for the UI. That is great. But, I am having a bug related to this new policy format that the PAP is returning.

Also, this seems to be the same format returned by the pap when a policy for a particular field is set or deleted.

I just realized that since the latest version I am getting a bug during declassification.  I am calling the PEP.declassify like this:

```
1) user information 
{"user_name":"agile","auth_type":"agile-local","role":"admin","password":"$2a$10$s4luwZ/wIQ1PlZuIfC3DueiMaY3qN9V9zw1xo6KYktUQvfFiMbAdm","credentials":{"xively":{"xivelymaster":"NU9grueAtYdQE0L7DdFlID3NBZuQn7tyyNvjXUvqoQnJ2rox","xivelyproduct":"Y1o-jUXIj66T1Ekb_Tjx","xivelysecret":"067d1c0ad522fa0315782888b4cf89741b0369ec"}},"id":"agile!@!agile-local","type":"/user","owner":"agile!@!agile-local"}

2) full record for the entity
pap.getEntityPolicyRecord


{"d":{"0":{"p":[{"op":"write"},{"op":"read"}],"c":1},"1":{"p":[{"op":"read"},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"attrEq","args":["role","admin"]}]}],"c":3},"2":{"p":[{"op":"read","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]},{"op":"read","locks":[{"lock":"hasType","args":["/user"]},{"lock":"attrEq","args":["role","admin"]}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"attrEq","args":["role","admin"]}]}],"c":1},"3":{"p":[{"op":"read"},{"op":"read","locks":[{"lock":"hasType","args":["/user"]},{"lock":"attrEq","args":["role","admin"]}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"attrEq","args":["role","admin"]}]}],"c":1},"4":{"p":[{"op":"read"}],"c":1},"5":{"p":[{"op":"read","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"attrEq","args":["role","admin"]}]}],"c":1},"6":{"p":[{"op":"read"},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"attrEq","args":["role","admin"]}]}],"c":2},"7":{"p":[{"op":"read","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]}],"c":1},"8":{"p":[{"op":"read"},{"op":"write","locks":[{"lock":"hasType","args":["/user"]},{"lock":"isOwner"}]}],"c":1}},"e":0,"o":{"s":1,"m":null,"p":{"actions":{"s":2,"m":null,"p":{}},"policies":{"s":3,"m":null,"p":{"policies":{"s":4,"m":null,"p":{}},"role":{"s":6,"m":null,"p":{}},"credentials":{"s":8,"m":null,"p":{}}}},"groups":{"s":1,"m":null,"p":{}},"entities":{"s":1,"m":null,"p":{}},"password":{"s":5,"m":null,"p":{}},"role":{"s":6,"m":null,"p":{}},"credentials":{"s":7,"m":null,"p":{}}}}}

3)  entity being read (is the same user):

{"user_name":"agile","auth_type":"agile-local","role":"admin","password":"$2a$10$s4luwZ/wIQ1PlZuIfC3DueiMaY3qN9V9zw1xo6KYktUQvfFiMbAdm","credentials":{"xively":{"xivelymaster":"NU9grueAtYdQE0L7DdFlID3NBZuQn7tyyNvjXUvqoQnJ2rox","xivelyproduct":"Y1o-jUXIj66T1Ekb_Tjx","xivelysecret":"067d1c0ad522fa0315782888b4cf89741b0369ec"}},"id":"agile!@!agile-local","type":"/user","owner":"agile!@!agile-local","client_id":"AGILE-OSJS"}


4)   pap.get(id,"") //attributePolicy for empty attribute.

{"flows":{"read":{"flows":[{"op":"read"}]},"write":{"flows":[{"op":"write","locks":{"hasType":[{"lock":"hasType","args":["/user"],"not":false}],"isOwner":[{"lock":"isOwner","not":false}]}},{"op":"write","locks":{"hasType":[{"lock":"hasType","args":["/user"],"not":false}],"attrEq":[{"lock":"attrEq","args":["role","admin"],"not":false}]}}]}}}
```



And I get the following stack trace:


```
Unhandled rejection Error: Policy: Cannot construct policy without valid flow specifications.
agile-security    |     at new Policy (/opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/ulocks/Policy.js:111:15)
agile-security    |     at /opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/UPFROnt/pap/api.js:82:29
agile-security    |     at tryCatcher (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/util.js:16:23)
agile-security    |     at Promise._settlePromiseFromHandler (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:512:31)
agile-security    |     at Promise._settlePromise (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:569:18)
agile-security    |     at Promise._settlePromise0 (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:614:10)
agile-security    |     at Promise._settlePromises (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:693:18)
agile-security    |     at Async._drainQueue (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:133:16)
agile-security    |     at Async._drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:143:10)
agile-security    |     at Immediate.Async.drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:17:14)
agile-security    |     at runCallback (timers.js:672:20)
agile-security    |     at tryOnImmediate (timers.js:645:5)
agile-security    |     at processImmediate [as _immediateCallback] (timers.js:617:5)
```

If you compare these logs to the logs in the [initial issue](https://github.com/SEDARI/UPFROnt/issues/4) there has been a change on how the policy structure looks like. My guess is that this is related?

**Comment by nopbyte** [comment]
also, for what I can see from my logs, it seems there is an issue with the checkWrite in the pap in some cases. However, debugging this is harder for me because I had to chain sequentially the promises setting policies and executing calls on the pdp :(

But, in case it helps this is what I get from my logs:

```
Unhandled rejection TypeError: Cannot read property 'hasOwnProperty' of undefined
agile-security    |     at _getProperty (/opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/UPFROnt/pap/api.js:110:37)
agile-security    |     at /opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/UPFROnt/pap/api.js:79:34
agile-security    |     at tryCatcher (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/util.js:16:23)
agile-security    |     at Promise._settlePromiseFromHandler (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:512:31)
agile-security    |     at Promise._settlePromise (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:569:18)
agile-security    |     at Promise._settlePromise0 (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:614:10)
agile-security    |     at Promise._settlePromises (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:693:18)
agile-security    |     at Async._drainQueue (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:133:16)
agile-security    |     at Async._drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:143:10)
agile-security    |     at Immediate.Async.drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:17:14)
agile-security    |     at runCallback (timers.js:672:20)
agile-security    |     at tryOnImmediate (timers.js:645:5)
agile-security    |     at processImmediate [as _immediateCallback] (timers.js:617:5)
agile-security    | Unhandled rejection Error: Policy: Cannot construct policy without valid flow specifications.
agile-security    |     at new Policy (/opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/ulocks/Policy.js:111:15)
agile-security    |     at /opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/UPFROnt/pap/api.js:82:29
agile-security    |     at tryCatcher (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/util.js:16:23)
agile-security    |     at Promise._settlePromiseFromHandler (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:512:31)
agile-security    |     at Promise._settlePromise (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:569:18)
agile-security    |     at Promise._settlePromise0 (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:614:10)
agile-security    |     at Promise._settlePromises (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:693:18)
agile-security    |     at Async._drainQueue (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:133:16)
agile-security    |     at Async._drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:143:10)
agile-security    |     at Immediate.Async.drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:17:14)
agile-security    |     at runCallback (timers.js:672:20)
agile-security    |     at tryOnImmediate (timers.js:645:5)
agile-security    |     at processImmediate [as _immediateCallback] (timers.js:617:5)
agile-security    | Unhandled rejection TypeError: Cannot read property 'hasOwnProperty' of undefined
agile-security    |     at _getProperty (/opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/UPFROnt/pap/api.js:110:37)
agile-security    |     at /opt/agile-idm-web-ui/node_modules/agile-policies/node_modules/UPFROnt/pap/api.js:79:34
agile-security    |     at tryCatcher (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/util.js:16:23)
agile-security    |     at Promise._settlePromiseFromHandler (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:512:31)
agile-security    |     at Promise._settlePromise (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:569:18)
agile-security    |     at Promise._settlePromise0 (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:614:10)
agile-security    |     at Promise._settlePromises (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/promise.js:693:18)
agile-security    |     at Async._drainQueue (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:133:16)
agile-security    |     at Async._drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:143:10)
agile-security    |     at Immediate.Async.drainQueues (/opt/agile-idm-web-ui/node_modules/bluebird/js/release/async.js:17:14)
agile-security    |     at runCallback (timers.js:672:20)
agile-security    |     at tryOnImmediate (timers.js:645:5)
agile-security    |     at processImmediate [as _immediateCallback] (timers.js:617:5)
```

**Comment by nopbyte** [comment]
hi @dschrecking 

I have executed the declassification on your master, and as you indicated it runs properly. Thanks for your help. I have checked further, and it is a problem on my side :(

The problem was that there were two versions due to some npm linking in my setup. And sometimes agile-security was taking then newer or the older versions of UPFROnt, so this is why some things worked and some other calls were braking when attempting to build the policy.

I am very sorry that I reported something that was not really a bug on your side!
